pkg/handlers/docker/mutation/customimage/inject.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/patches"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/variables"
+	"github.com/d2iq-labs/capi-runtime-extensions/pkg/handlers/generic/clusterconfig"
+)
+
+const (
+	// HandlerNamePatch is the name of the inject handler.
+	HandlerNamePatch = "DockerCustomImagePatch"
+
+	defaultKinDImageRepository = "ghcr.io/mesosphere/kind-node"
+)
+
+type customImagePatchHandler struct {
+	variableName      string
+	variableFieldPath []string
+}
+
+var (
+	_ commonhandlers.Named     = &customImagePatchHandler{}
+	_ mutation.GeneratePatches = &customImagePatchHandler{}
+	_ mutation.MetaMutater     = &customImagePatchHandler{}
+)
+
+func NewPatch() *customImagePatchHandler {
+	return newCustomImagePatchHandler(variableName)
+}
+
+func NewMetaPatch() *customImagePatchHandler {
+	return newCustomImagePatchHandler(clusterconfig.MetaVariableName, variableName)
+}
+
+func newCustomImagePatchHandler(
+	variableName string,
+	variableFieldPath ...string,
+) *customImagePatchHandler {
+	return &customImagePatchHandler{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *customImagePatchHandler) Name() string {
+	return HandlerNamePatch
+}
+
+func (h *customImagePatchHandler) Mutate(
+	ctx context.Context,
+	obj runtime.Object,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	_ client.ObjectKey,
+) error {
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"holderRef", holderRef,
+	)
+
+	customImageVar, found, err := variables.Get[v1alpha1.OCIImage](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		return err
+	}
+	if !found {
+		log.V(5).Info("Docker customImage variable not defined, using default KinD node image")
+	}
+
+	log = log.WithValues(
+		"variableName",
+		h.variableName,
+		"variableFieldPath",
+		h.variableFieldPath,
+		"variableValue",
+		customImageVar,
+	)
+
+	err = patches.Generate(
+		obj,
+		vars,
+		&holderRef,
+		selectors.InfrastructureWorkerMachineTemplates("v1beta1", "DockerMachineTemplate"),
+		log,
+		func(obj *unstructured.Unstructured) error {
+			fieldPath := []string{"builtin", "machineDeployment", "version"}
+
+			if customImageVar == "" {
+				kubernetesVersion, found, err := variables.Get[string](
+					vars,
+					fieldPath[0],
+					fieldPath[1:]...)
+				if err != nil {
+					return err
+				}
+				if !found {
+					return fmt.Errorf(
+						"missing required builtin variable: %s",
+						strings.Join(fieldPath, "."),
+					)
+				}
+
+				customImageVar = v1alpha1.OCIImage(
+					defaultKinDImageRepository + ":" + kubernetesVersion,
+				)
+			}
+
+			log.WithValues(
+				"patchedObjectKind", obj.GetObjectKind().GroupVersionKind().String(),
+				"patchedObjectName", client.ObjectKeyFromObject(obj),
```

Comment:
Can we log the resolved custom image var?

Reply:
Done.